File: src/games/sudoku/gameplay/analysis/techniqueModules/xWing.ts
import { sudokuDigits } from '../../../types';
import { columnCellIndexes, rowCellIndexes } from '../bitmask';
import { buildCandidateEliminationMove, getHouseDigitMatches } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';

export const xWingTechnique: SudokuTechniqueDispatcher = {
  technique: 'x-wing',
  tier: 'hard',
  findMove(state) {
    for (const digit of sudokuDigits) {
      const rowCandidates = rowCellIndexes.map((cells, row) => ({
        row,
        cols: getHouseDigitMatches(state, cells, digit).map((index) => index % 9),
      })).filter(({ cols }) => cols.length === 2);

      for (let first = 0; first < rowCandidates.length; first += 1) {
        for (let second = first + 1; second < rowCandidates.length; second += 1) {
          const left = rowCandidates[first];
          const right = rowCandidates[second];
          if (left.cols[0] !== right.cols[0] || left.cols[1] !== right.cols[1]) {
            continue;
          }

          const move = buildCandidateEliminationMove({
            technique: 'x-wing',
            eliminations: left.cols.flatMap((col) => columnCellIndexes[col]
              .filter((index) => Math.floor(index / 9) !== left.row && Math.floor(index / 9) !== right.row)
              .filter((index) => state.board[index] === 0 && (state.candidateMask[index] & (1 << (digit - 1))) !== 0)
              .map((index) => ({ index, digit }))),
            evidenceCells: [
              (left.row * 9) + left.cols[0],
              (left.row * 9) + left.cols[1],
              (right.row * 9) + right.cols[0],
              (right.row * 9) + right.cols[1],
            ],
            houses: [
              { kind: 'row', index: left.row },
              { kind: 'row', index: right.row },
              { kind: 'column', index: left.cols[0] },
              { kind: 'column', index: left.cols[1] },
            ],
          });
          if (move) {
            return move;
          }
        }
      }

      const columnCandidates = columnCellIndexes.map((cells, col) => ({
        col,
        rows: getHouseDigitMatches(state, cells, digit).map((index) => Math.floor(index / 9)),
      })).filter(({ rows }) => rows.length === 2);

      for (let first = 0; first < columnCandidates.length; first += 1) {
        for (let second = first + 1; second < columnCandidates.length; second += 1) {
          const left = columnCandidates[first];
          const right = columnCandidates[second];
          if (left.rows[0] !== right.rows[0] || left.rows[1] !== right.rows[1]) {
            continue;
          }

          const move = buildCandidateEliminationMove({
            technique: 'x-wing',
            eliminations: left.rows.flatMap((row) => rowCellIndexes[row]
              .filter((index) => (index % 9) !== left.col && (index % 9) !== right.col)
              .filter((index) => state.board[index] === 0 && (state.candidateMask[index] & (1 << (digit - 1))) !== 0)
              .map((index) => ({ index, digit }))),
            evidenceCells: [
              (left.rows[0] * 9) + left.col,
              (left.rows[1] * 9) + left.col,
              (right.rows[0] * 9) + right.col,
              (right.rows[1] * 9) + right.col,
            ],
            houses: [
              { kind: 'column', index: left.col },
              { kind: 'column', index: right.col },
              { kind: 'row', index: left.rows[0] },
              { kind: 'row', index: left.rows[1] },
            ],
          });
          if (move) {
            return move;
          }
        }
      }
    }

    return null;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/XWing.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*

object XWingTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.X_WING
    override val tier = Difficulty.HARD

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        var best: SudokuCanonicalMove? = null
        var bestComplexity = Int.MAX_VALUE

        for (digit in 1..9) {
            val rowCandidates = rowCellIndexes.mapIndexed { row, cells -> row to getHouseDigitMatches(state, cells, digit).map { it % 9 } }
                .filter { it.second.size == 2 }
            val rowComplexity = rowCandidates.size
            if (rowComplexity < bestComplexity) {
                outer@ for (first in rowCandidates.indices) {
                    for (second in (first + 1) until rowCandidates.size) {
                        val (leftRow, leftCols) = rowCandidates[first]
                        val (rightRow, rightCols) = rowCandidates[second]
                        if (leftCols[0] != rightCols[0] || leftCols[1] != rightCols[1]) continue
                        val eliminations = leftCols.flatMap { col ->
                            columnCellIndexes[col].filter { it / 9 != leftRow && it / 9 != rightRow }
                                .filter { state.board[it] == 0 && (state.candidateMask[it] and digitToBit[digit]) != 0 }
                                .map { it to digit }
                        }
                        val move = buildCandidateEliminationMove(
                            SudokuTechnique.X_WING, eliminations,
                            listOf(leftRow * 9 + leftCols[0], leftRow * 9 + leftCols[1], rightRow * 9 + rightCols[0], rightRow * 9 + rightCols[1]),
                            listOf(SudokuHouseRef("row", leftRow), SudokuHouseRef("row", rightRow), SudokuHouseRef("column", leftCols[0]), SudokuHouseRef("column", leftCols[1])),
                            rowComplexity,
                        )
                        if (move != null) { best = move; bestComplexity = rowComplexity; break@outer }
                    }
                }
            }

            val columnCandidates = columnCellIndexes.mapIndexed { col, cells -> col to getHouseDigitMatches(state, cells, digit).map { it / 9 } }
                .filter { it.second.size == 2 }
            val colComplexity = columnCandidates.size
            if (colComplexity < bestComplexity) {
                outer@ for (first in columnCandidates.indices) {
                    for (second in (first + 1) until columnCandidates.size) {
                        val (leftCol, leftRows) = columnCandidates[first]
                        val (rightCol, rightRows) = columnCandidates[second]
                        if (leftRows[0] != rightRows[0] || leftRows[1] != rightRows[1]) continue
                        val eliminations = leftRows.flatMap { row ->
                            rowCellIndexes[row].filter { it % 9 != leftCol && it % 9 != rightCol }
                                .filter { state.board[it] == 0 && (state.candidateMask[it] and digitToBit[digit]) != 0 }
                                .map { it to digit }
                        }
                        val move = buildCandidateEliminationMove(
                            SudokuTechnique.X_WING, eliminations,
                            listOf(leftRows[0] * 9 + leftCol, leftRows[1] * 9 + leftCol, rightRows[0] * 9 + rightCol, rightRows[1] * 9 + rightCol),
                            listOf(SudokuHouseRef("column", leftCol), SudokuHouseRef("column", rightCol), SudokuHouseRef("row", leftRows[0]), SudokuHouseRef("row", leftRows[1])),
                            colComplexity,
                        )
                        if (move != null) { best = move; bestComplexity = colComplexity; break@outer }
                    }
                }
            }
        }
        return best
    }
}
